
#' generate.peaks.from.gmm is a function that takes the output of a GMM and generates peaks
#'
#' @param dp_data Results from fitting a Dirichlet Process GMM model
#' @param PARAMETERS A PARAMETERS list with the parameters indicated in the DPDE4PM function
#' @param GENEINFO A list of gene attributes generated by the get.gene.anno function
#'
#' @return
#' \describe{
#'  \item{merged.peaks.filtered.rna}{A merged peaks GenomicRanges object in RNA coordinates}
#'  \item{merged.peaks.filtered.genome}{A merged peaks GenomicRanges object in genomic coordinates}
#' }
.generate.peaks.from.gmm = function(
  dp_data,
  PARAMETERS,
  GENEINFO
){

  # Filtering by Threshold
  dp_data = dp_data[dp_data$Weights > PARAMETERS$WEIGHT.THRESHOLD,]
  if(nrow(dp_data) == 0){
    warning("No Peaks Survive Past The Threshold. Consider Lowering Requirements or Tuning Parameters.",
            call. = TRUE, domain = NULL)
    return(list(GenomicRanges::GRanges(),  GenomicRanges::GRanges()))
  }
  dp_data = dp_data[order(-dp_data$Weights),]

  # Creating Peaks
  merged.peaks = data.frame(
    "chr" = GENEINFO$chr,
    "start" = round(dp_data$Mu - PARAMETERS$N.SD*dp_data$Sigma),
    "end" = round(dp_data$Mu + PARAMETERS$N.SD*dp_data$Sigma),
    "name" = GENEINFO$gene,
    "strand" = GENEINFO$strand,
    "weights" = dp_data$Weights,
    "i" = dp_data$i,
    "j" = seq(1, nrow(dp_data), 1),
    stringsAsFactors = F
  )

  # Filtering Overlapping Peaks
  merged.peaks.gr =  GenomicRanges::makeGRangesFromDataFrame(merged.peaks, keep.extra.columns = T)
  merged.peaks.filtered.rna = do.call(c, lapply(1:length(merged.peaks.gr), function(i){
    if(i == 1){
      tmp = merged.peaks.gr[1]
    }else{
      tmp = GenomicRanges::setdiff(merged.peaks.gr[i], merged.peaks.gr[1:(i-1)])
      if(length(tmp) > 0){S4Vectors::mcols(tmp) = S4Vectors::mcols(merged.peaks.gr[i])}
    }
    tmp
  }))

  # Transferring to Genomic Coordinates
  merged.peaks.genome = merged.peaks.filtered.rna
  GenomicRanges::end(merged.peaks.genome) = GENEINFO$RNA2DNA[GenomicRanges::end(merged.peaks.genome)]
  GenomicRanges::start(merged.peaks.genome) = GENEINFO$RNA2DNA[GenomicRanges::start(merged.peaks.genome)]
  anno_gr = GenomicRanges::makeGRangesFromDataFrame(GENEINFO$anno)

  # Filtering Out Introns
  merged.peaks.filtered.genome = do.call(c, lapply(1:length(merged.peaks.genome), function(i){
      tmp = GenomicRanges::intersect(merged.peaks.genome[i], anno_gr)
      if(length(tmp) > 0){S4Vectors::mcols(tmp) = S4Vectors::mcols(merged.peaks.genome[i])}
      tmp
  }))

  return(list(merged.peaks.filtered.rna, merged.peaks.filtered.genome))
}
